File: leecode/牛客/剑指Offer/Binary_tree_KthNode.go
//package main

/*
@author: forgotten_liu
@projectName: Golang_study
@file: Binary_tree_KthNode
@time: 2020/12/28 13:49
@IDE: GoLand
@desc: 二叉树的第 n 个节点
给定一棵二叉搜索树，请找出其中的第k小的结点。
示例1
输入
{5,3,7,2,4,6,8},3
返回值
{4}
     5
  3    7
 2 4  6 8
*/

package main

import "fmt"

type TreeNode struct {
	Val int
  	Left *TreeNode
  	Right *TreeNode
}


/**
 *
 * @param pRoot TreeNode类
 * @param k int整型
 * @return TreeNode类
 */

// 由于返回值需要为 *TreeNode 类型 ,所以 无法使用 切片存入遍历值,需要用切片存入遍历节点
func KthNode( pRoot *TreeNode ,  k int ) *TreeNode {
	if pRoot == nil || k <= 0 {
		return nil
	}
	arr := dfs(pRoot)
	if len(arr) < k {
		return nil
	}
	return arr[k-1]
}

// 中序遍历树
func dfs(Root *TreeNode) (arr []*TreeNode) {
	if Root == nil {
		return nil
	}
	// 在 dfs(Root.Left) []*TreeNode 内开始存入 节点值
	arr = append(dfs(Root.Left), Root)
	arr = append(arr, dfs(Root.Right)...)
	return arr
}

func main() {
	leftLeft := TreeNode{
		Val: 2,
	}
	leftRight := TreeNode{
		Val: 4,
	}
	left := TreeNode{
		Val:   3,
		Left:  &leftLeft,
		Right: &leftRight,
	}
	rightLeft := TreeNode{
		Val: 6,
	}
	rightRight := TreeNode{
		Val: 8,
	}
	right := TreeNode{
		Val: 7,
		Left: &rightLeft,
		Right: &rightRight,
	}
	tree := TreeNode{
		Val: 5,
		Left: &left,
		Right: &right,
	}
	fmt.Println(*KthNode(&tree, 3))
}

File: leecode/牛客/剑指Offer/Binary_tree_KthNode.py
#!/usr/bin/python3

# -*- coding:utf-8 -*-
"""

@author: forgotten_liu
@projectName: python_study
@file: Binary_tree_KthNode
@time: 2020/12/28 9:49
@IDE: PyCharm
@desc: 二叉树的第 n 个节点
给定一棵二叉搜索树，请找出其中的第k小的结点。
示例1
输入
{5,3,7,2,4,6,8},3
返回值
{4}
     5
  3    7
 2 4  6 8
"""


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


# 方法一
class Solution:
    # 返回对应节点TreeNode
    index = 0

    def KthNode(self, pRoot, k):
        # write code here
        if not pRoot:
            return None
        left = self.KthNode(pRoot.left, k)
        self.index += 1
        if self.index == k:
            return pRoot
        right = self.KthNode(pRoot.right, k)
        if left:
            return left
        elif right:
            return right


# 方法二
class Solution1:
    # 返回对应节点TreeNode

    def KthNode(self, pRoot, k):
        self.arr = []
        if not pRoot:
            return None
        self.tree_to_list(pRoot)
        if 0 < k <= len(self.arr):
            return self.arr[k-1]

    # 中序遍历 得到的列表是有序的
    def tree_to_list(self, Root):
        if not Root:
            return None
        self.tree_to_list(Root.left)
        self.arr.append(Root)
        self.tree_to_list(Root.right)


tree = TreeNode(5)
tree.left = TreeNode(3)
tree.right = TreeNode(7)
tree.left.left = TreeNode(2)
tree.left.right = TreeNode(4)
tree.right.left = TreeNode(6)
tree.right.right = TreeNode(8)
s = Solution1()
print(s.KthNode(tree, 1))